### PR TITLE
Add support for custom shoot fields

### LINKED ISSUE #686: Define Custom Labels for Shoot Overview
Define custom labels / annotations for the dashboard that are rendered as optional column in the shoot overview. This feature can help organizing clusters.

### PR DESCRIPTION
**Pull request by petersutter**
**What this PR does / why we need it**:
This PR adds support for custom shoot fields, that can be defined per project by specifying `metadata.annotations["dashboard.gardener.cloud/customFieldsShoot"]`.
The fields can be configured to be displayed on the cluster list and cluster details page.
Custom fields do not show up on the `ALL_PROJECTS` page.

Screenshot 2020-12-02 at 15 55 00

Screenshot 2020-12-02 at 15 55 46

Screenshot 2020-12-02 at 15 56 47

Screenshot 2020-12-02 at 15 57 15

| Property | Type | Default | Required | Description |
|---|---|---|---|---|
| name | String | | ✔️ | Name of the custom field |
| path | String | | ✔️ | Path in shoot resource, of which the value must be of primitive type (no object / array). Use [lodash get](https://lodash.com/docs/4.17.15#get) path syntax, e.g. `metadata.labels["shoot.gardener.cloud/status"]` or `spec.networking.type` |
| icon | String | | | MDI icon for field on the cluster details page. See https://materialdesignicons.com/ for available icons. Must be in the format: `mdi-<icon-name>`. |
| tooltip | String | | | Tooltip for the custom field that appears when hovering with the mouse over the value  |
| defaultValue | String/Number | | | Default value, in case there is no value for the given `path` |
| showColumn | Bool | true | | Field shall appear as column in the cluster list |
| columnSelectedByDefault | Bool | true | | Indicates if field shall be selected by default on the cluster list (not hidden by default) |
| weight | Number | 0 | | Defines the order of the column. The standard columns start with weight 100 and continue in 100 increments (200, 300, ..) |
| sortable | Bool | true | | Indicates if column is sortable on the cluster list.  |
| searchable | Bool | true | | Field shall appear in a dedicated card (`Custom Fields`) on the cluster details page |
| showDetails | Bool | true | | Indicates if field shall appear in a dedicated card (`Custom Fields`) on the cluster details page |

Example project yaml:
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Project
metadata:
  annotations:
    dashboard.gardener.cloud/shootCustomFields: |
      {
        "shootStatus": {
          "name": "Shoot Status",
          "path": "metadata.labels[\"shoot.gardener.cloud/status\"]",
          "icon": "mdi-heart-pulse",
          "tooltip": "Indicates the health status of the cluster"
          "defaultValue": "unknown",
          "showColumn": true,
          "columnSelectedByDefault": true,
          "weight": 950,
          "searchable": true,
          "sortable": true,
          "showDetails": true
        },
        "networking": {
          "name": "Networking Type",
          "path": "spec.networking.type",
          "icon": "mdi-table-network",
          "showColumn": false
        }
      }
```

**Which issue(s) this PR fixes**:
Fixes #686 

**Special notes for your reviewer**:

**Release note**:

```feature user
You can now define your own (project specific) custom fields for your shoots.
- The custom fields can be displayed on the cluster list and / or on the cluster details page
- The defined custom fields are listed on the `ADMINISTRATION` page
- You have to set `metadata.annotations["dashboard.gardener.cloud/customFieldsShoot"]`on your project with your custom field definitions, e.g. using `kubectl`. See [#890](https://github.com/gardener/dashboard/pull/890) for more details
```

### CODE REVIEW

File: frontend/src/components/CustomField.vue [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    </v-list>
+  </g-popper>
+</template>
+
+<script>
+
+import GPopper from '@/components/GPopper'
+
+export default {
+  name: 'worker-group',
```
Comment (holgerkoser) [inline frontend/src/components/CustomField.vue:124]:
```suggestion
  name: 'custom-field',
```

File: frontend/src/components/CustomField.vue [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    },
+    searchable: {
+      type: Boolean
+    },
+    showDetails: {
+      type: Boolean
+    }
+  },
+  computed: {
+  }
```
Comment (holgerkoser) [inline frontend/src/components/CustomField.vue:164]:
```suggestion
```

File: frontend/src/components/CustomField.vue [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+<script>
+
+import GPopper from '@/components/GPopper'
+
+export default {
+  name: 'worker-group',
+  components: {
+    GPopper
+  },
+  props: {
```
Comment (holgerkoser) [inline frontend/src/components/CustomField.vue:128]: No required properties and no default values?

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  <v-card v-if="customFields">
+    <v-toolbar flat dark dense color="cyan darken-2">
+      <v-toolbar-title class="subtitle-1">Custom Fields</v-toolbar-title>
+    </v-toolbar>
+    <v-list>
+      <template v-for="(customField, index) in customFields" >
+        <v-divider v-if="index !== 0" inset :key="`${customField.key}-divider`"></v-divider>
+        <v-list-item :key="customField.key">
+          <v-list-item-icon>
+            <v-icon color="cyan darken-2" v-if="customField.icon">{{customField.icon}}</v-icon>
```
Comment (holgerkoser) [inline frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:17]: Shouldn't the v-if be in the `<v-list-item-icon>` tag?

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+<script>
+import { mapGetters } from 'vuex'
+import filter from 'lodash/filter'
+import get from 'lodash/get'
+import map from 'lodash/map'
+
+import { shootItem } from '@/mixins/shootItem'
+
+export default {
```
Comment (holgerkoser) [inline frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:45]:
```suggestion
export default {
  name: 'shoot-custom-fields-card',
```

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue (hunk 4 above)
Comment (holgerkoser) [reply frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:17]: The name of the first component was `CustomField`. Why is this component named  `ShootCustomFieldsCard` and not `CustomFieldsCard`

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  props: {
+    shootItem: {
```
Comment (holgerkoser) [inline frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:47]: This property should be defined in the mixin and not by the user of the mixin. The mixin is not self contained.

Comment (holgerkoser) [reply frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:47]: I could be better to pass the custom fields as property.

File: frontend/src/components/ShootListRow.vue [hunk 7]
```diff
@@ -6,119 +6,137 @@ SPDX-License-Identifier: Apache-2.0
...
-    <td v-if="this.headerVisible['accessRestrictions']">
-      <access-restriction-chips :selectedAccessRestrictions="shootSelectedAccessRestrictions"></access-restriction-chips>
-    </td>
-    <td class="nowrap" v-if="this.headerVisible['ticket']">
-      <v-tooltip top>
-        <template v-slot:activator="{ on }">
-          <div v-on="on">
+    <td v-for="cell in cells" :key="cell.header.value" :class="cell.header.class">
+      <template v-if="cell.header.value === 'project'">
+        <router-link class="cyan--text text--darken-2" :to="{ name: 'ShootList', params: { namespace: shootNamespace } }">
```
Comment (holgerkoser) [inline frontend/src/components/ShootListRow.vue:11]: Not part of this PR but I want to mention it. The name prefix `shoot` make things very complicated. I am not sure if this is really necessary.

File: frontend/src/router/guards.js [hunk 8]
```diff
@@ -135,7 +135,7 @@ function ensureDataLoaded (store, localStorage) {
             deactivatedReconciliation: isAdmin,
             hideTicketsWithLabel: isAdmin
           }
-          const shootListFilters = defaults(localStorage.getObject('shootListFilter'), defaultFilter)
+          const shootListFilters = defaults(localStorage.getObject('shootList_filter'), defaultFilter)
```
Comment (holgerkoser) [inline frontend/src/router/guards.js:138]: What is the reason behind this kind of renaming?

File: frontend/src/store/index.js [hunk 9]
```diff
@@ -527,6 +532,49 @@ const getters = {
...
+      customFieldsShoot = JSON.parse(customFieldsShoot)
+    } catch (error) {
+      console.error('could not parse custom fields', error.message)
+      return undefined
+    }
+
+    customFieldsShoot = pickBy(customFieldsShoot, customFields => {
+      return !isEmpty(get(customFields, 'path')) &&
+        !isEmpty(get(customFields, 'name'))
+    })
```
Comment (holgerkoser) [inline frontend/src/store/index.js:559]:
```suggestion
    customFieldsShoot = pickBy(customFieldsShoot, ({ path, name }) => name && path)
```

File: frontend/src/store/index.js [hunk 10]
```diff
@@ -527,6 +532,49 @@ const getters = {
...
+    let customFieldsShoot = get(getters.projectFromProjectList, 'metadata.annotations["dashboard.gardener.cloud/customFieldsShoot"]')
+    if (!customFieldsShoot) {
+      return undefined
+    }
+
+    try {
+      customFieldsShoot = JSON.parse(customFieldsShoot)
+    } catch (error) {
+      console.error('could not parse custom fields', error.message)
+      return undefined
```
Comment (holgerkoser) [inline frontend/src/store/index.js:553]:
```suggestion
      return
```

File: frontend/src/store/index.js [hunk 11]
```diff
@@ -527,6 +532,49 @@ const getters = {
...
+      return {
+        ...customFields,
+        key
+      }
+    })
+  },
+  customFieldsShoot (state, getters) {
+    let customFieldsShoot = get(getters.projectFromProjectList, 'metadata.annotations["dashboard.gardener.cloud/customFieldsShoot"]')
+    if (!customFieldsShoot) {
+      return undefined
```
Comment (holgerkoser) [inline frontend/src/store/index.js:546]:
```suggestion
      return
```

File: frontend/src/store/index.js [hunk 12]
```diff
@@ -527,6 +532,49 @@ const getters = {
...
+        showColumn: true,
+        columnSelectedByDefault: true,
+        showDetails: true,
+        sortable: true,
+        searchable: true
+      }
+      return defaults({
+        ...customFields,
+        defaultValue
+      }, defaultProperties)
```
Comment (holgerkoser) [inline frontend/src/store/index.js:574]:
```suggestion
      return {
        ...defaultProperties,
        ...customFields,
        defaultValue
      }
```

File: frontend/src/store/index.js [hunk 13]
```diff
@@ -527,6 +532,49 @@ const getters = {
...
+
+    customFieldsShoot = pickBy(customFieldsShoot, customFields => {
+      return !isEmpty(get(customFields, 'path')) &&
+        !isEmpty(get(customFields, 'name'))
+    })
+
+    customFieldsShoot = mapKeys(customFieldsShoot, (customFields, key) => `Z_${key}`)
+    customFieldsShoot = mapValues(customFieldsShoot, customFields => {
+      const defaultValue = !isObject(customFields.defaultValue) ? customFields.defaultValue : undefined
+      const defaultProperties = {
```
Comment (holgerkoser) [inline frontend/src/store/index.js:564]: move outside of mapValues

File: frontend/src/store/modules/shoots.js [hunk 14]
```diff
@@ -503,41 +508,45 @@ const setSortedItems = (state, rootState) => {
...
-  let items = state.sortedShoots
-  if (state.searchValue) {
-    const predicate = item => {
-      let found = true
-      forEach(state.searchValue, value => {
-        if (includes(getRawVal(item, 'name'), value)) {
-          return
+function setFilteredAndSortedItems (state, rootState) {
+  function matchesShoot (searchValue) {
+    const searchableCustomFields = filter(store.getters.customFieldsListShoot, ['searchable', true])
```
Comment (holgerkoser) [inline frontend/src/store/modules/shoots.js:513]: we don't have the store in the arguments?

Comment (holgerkoser) [reply frontend/src/store/modules/shoots.js:513]: I would be better to pass getters to this function

File: frontend/src/views/Administration.vue [hunk 15]
```diff
@@ -149,6 +149,48 @@ SPDX-License-Identifier: Apache-2.0
                     </v-list-item-content>
                   </v-list-item>
                 </template>
+                <template v-if="customFieldsListShoot">
```
Comment (holgerkoser) [inline frontend/src/views/Administration.vue:152]: The name should be `customShootFieldList` or `customShootFields`

File: frontend/src/views/Administration.vue [hunk 16]
```diff
@@ -149,6 +149,48 @@ SPDX-License-Identifier: Apache-2.0
...
+                          :icon="icon"
+                          :tooltip="tooltip"
+                          :default-value="defaultValue"
+                          :show-column="showColumn"
+                          :weight="weight"
+                          :column-selected-by-default="columnSelectedByDefault"
+                          :show-details="showDetails"
+                          :searchable="searchable"
+                        ></custom-field>
+                        <span v-if="!customFieldsListShoot || customFieldsListShoot.length === 0" class="font-weight-light text--disabled">Not defined</span>
```
Comment (holgerkoser) [inline frontend/src/views/Administration.vue:189]:
```suggestion
                        <span v-if="!customFieldsListShoot || !customFieldsListShoot.length" class="font-weight-light text--disabled">Not defined</span>
```

File: frontend/src/views/ShootList.vue [hunk 17]
```diff
@@ -189,48 +218,33 @@ export default {
...
   watch: {
     options (value) {
-      if (value) {
-        this.$localStorage.setObject('dataTable_options', pick(value, ['sortBy', 'sortDesc', 'itemsPerPage']))
-        this.setShootListSortParams(value)
+      if (!value) {
+        return
       }
+      let tableOptions = pick(value, ['sortBy', 'sortDesc', 'itemsPerPage'])
+      if (startsWith(tableOptions.sortBy, 'Z_')) {
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:235]: I think we need a constant CUSTOM_FIELD_PREFIX for `Z_`

File: frontend/src/views/ShootList.vue [hunk 18]
```diff
@@ -367,11 +363,81 @@ export default {
...
+        }
+      })
+    },
+    allHeaders () {
+      const allHeaders = [...this.standardHeaders, ...this.customHeaders]
+      return sortBy(allHeaders, ['weight', 'text'])
+    },
     headers () {
-      return this.allHeaders.filter(e => e.hidden === false)
+      return filter(this.allHeaders, ['hidden', false])
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:437]: 👍

File: frontend/src/views/ShootList.vue [hunk 19]
```diff
@@ -367,11 +363,81 @@ export default {
...
+        columnSelectedByDefault: defaultChecked,
+        tooltip,
+        defaultValue,
+        sortable,
+        weight
+      }, index) => {
+        return {
+          customField: true,
+          text: upperCase(name),
+          value: key,
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:419]: Why is this named value if it is the key or name of the custom field?

File: frontend/src/views/ShootList.vue [hunk 20]
```diff
@@ -171,14 +188,26 @@ SPDX-License-Identifier: Apache-2.0
...
+import zipObject from 'lodash/zipObject'
 import ShootListRow from '@/components/ShootListRow'
 const ShootAccessCard = () => import('@/components/ShootDetails/ShootAccessCard')
 
+function mapHeader (headers, key) {
+  const keys = map(headers, 'value')
+  const checkedValues = map(headers, key)
+  return zipObject(keys, checkedValues)
+}
+
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:210]:
```suggestion
function mapHeader (headers, valueKey) {
  const obj = {}
  for (const { value: key, [valueKey]: value } of headers) {
    obj[key] = value
  }
  return obj
}

```

File: frontend/src/views/ShootList.vue (hunk 17 above)
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:235]: I don't understand what is happening here!?

File: frontend/src/views/ShootList.vue [hunk 21]
```diff
@@ -189,48 +218,33 @@ export default {
...
   },
   watch: {
     options (value) {
-      if (value) {
-        this.$localStorage.setObject('dataTable_options', pick(value, ['sortBy', 'sortDesc', 'itemsPerPage']))
-        this.setShootListSortParams(value)
+      if (!value) {
+        return
       }
+      let tableOptions = pick(value, ['sortBy', 'sortDesc', 'itemsPerPage'])
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:234]: Whenever we use let for an object I assume something is wrong

File: frontend/src/views/ShootList.vue [hunk 22]
```diff
@@ -265,58 +279,37 @@ export default {
...
-      const checkedColumns = this.$localStorage.getObject('dataTable_checkedColumns') || {}
-      for (const header of this.allHeaders) {
-        header.checked = get(checkedColumns, header.value, header.defaultChecked)
-      }
+    updateTableSettings () {
+      this.checkedColumns = { ...this.$localStorage.getObject('shootList_checkedColumns'), ...this.$localStorage.getObject(`shootList_checkedColumns_${this.projectName}`) }
+
+      const projectSpecificTableOptions = this.$localStorage.getObject(`shootList_options_${this.projectName}`)
+      const tableOptions = this.$localStorage.getObject('shootList_options')
+      this.options = defaults(projectSpecificTableOptions, tableOptions, this.defaultTableOptions)
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:304]:
```suggestion
      this.options = {
        ...this.defaultTableOptions,
        ...tableOptions,
        ...projectSpecificTableOptions
      }
```

File: frontend/src/views/ShootList.vue [hunk 23]
```diff
@@ -265,58 +279,37 @@ export default {
...
+      this.options = this.defaultTableOptions
     },
-    loadColumnsChecked () {
-      const checkedColumns = this.$localStorage.getObject('dataTable_checkedColumns') || {}
-      for (const header of this.allHeaders) {
-        header.checked = get(checkedColumns, header.value, header.defaultChecked)
-      }
+    updateTableSettings () {
+      this.checkedColumns = { ...this.$localStorage.getObject('shootList_checkedColumns'), ...this.$localStorage.getObject(`shootList_checkedColumns_${this.projectName}`) }
+
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:301]:
```suggestion
      const checkedColumns = ...this.$localStorage.getObject('shootList_checkedColumns')
      const projectSpecificCheckedColumns = this.$localStorage.getObject(`shootList_checkedColumns_${this.projectName}`) 
      this.checkedColumns = { 
        ...checkedColumns, 
        ...projectSpecificCheckedColumns
      }
```

File: frontend/src/views/ShootList.vue [hunk 24]
```diff
@@ -265,58 +279,37 @@ export default {
...
+      this.$localStorage.setObject('shootList_checkedColumns', this.currentStandardCheckedColumns)
+      if (isEmpty(this.currentCustomCheckedColumns)) {
+        this.$localStorage.removeItem(`shootList_checkedColumns_${this.projectName}`)
+      } else {
+        this.$localStorage.setObject(`shootList_checkedColumns_${this.projectName}`, this.currentCustomCheckedColumns)
       }
-      this.saveColumnsChecked()
+    },
+    resetTableSettings () {
+      this.checkedColumns = { ...this.defaultStandardCheckedColumns, ...this.defaultCustomCheckedColumns }
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:294]:
```suggestion
      this.checkedColumns = { 
        ...this.defaultStandardCheckedColumns, 
        ...this.defaultCustomCheckedColumns 
      }
```

File: frontend/src/views/ShootList.vue [hunk 25]
```diff
@@ -189,48 +218,33 @@ export default {
...
+      let tableOptions = pick(value, ['sortBy', 'sortDesc', 'itemsPerPage'])
+      if (startsWith(tableOptions.sortBy, 'Z_')) {
+        this.$localStorage.setObject(`shootList_options_${this.projectName}`, pick(tableOptions, ['sortBy', 'sortDesc']))
+
+        // only update itemsPerPage value
+        const itemsPerPage = tableOptions.itemsPerPage
+        tableOptions = defaults(this.$localStorage.getObject('shootList_options'), this.defaultTableOptions)
+        tableOptions.itemsPerPage = itemsPerPage
+      }
+      this.$localStorage.setObject('shootList_options', tableOptions)
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:243]:
```suggestion
  const { sortBy, sortDesc, itemsPerPage } = value
  if (startsWith(sortBy, 'Z_')) {
    this.$localStorage.setObject(`shootList_options_${this.projectName}`, { sortBy, sortDesc })
    const currentTableOptions = this.$localStorage.getObject('shootList_options')
    const tableOptions = {
      ...this.defaultTableOptions,
      ...currentTableOptions,
      itemsPerPage
    }
    this.$localStorage.setObject('shootList_options', tableOptions)
  } else {
    this.$localStorage.setObject(`shootList_options_${this.projectName}`, { sortBy, sortDesc, itemsPerPage })
  }
```

File: frontend/src/router/guards.js (hunk 8 above)
Comment (petersutter) [reply frontend/src/router/guards.js:138]: to be consistent with `shootList_options`

File: frontend/src/store/modules/shoots.js (hunk 14 above)
Comment (petersutter) [reply frontend/src/store/modules/shoots.js:513]: yes but didn't we agree to make these changes in the upcoming refactoring?

File: frontend/src/views/ShootList.vue (hunk 19 above)
Comment (petersutter) [reply frontend/src/views/ShootList.vue:419]: `value` is defined / required by vuetify for a header

File: frontend/src/components/ShootListRow.vue (hunk 7 above)
Comment (petersutter) [reply frontend/src/components/ShootListRow.vue:11]: this is coming from the mixin, but we could introduce a `namespace` computed property

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue (hunk 4 above)
Comment (petersutter) [reply frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:17]: > Shouldn't the v-if be in the `<v-list-item-icon>` tag?

then we would need to add a padding, otherwise it wouldn't be aligned anymore, that's why I haven't put the v-if on the `v-list-item-icon` tag

Comment (petersutter) [reply frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue:17]: > The name of the first component was `CustomField`. Why is this component named `ShootCustomFieldsCard` and not `CustomFieldsCard`

I think we should rather add a prefix to the CustomField component. Maybe we want to introduce custom fields for other resources, like projects etcs. That was also the main reason why I prefixed this component name

File: frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue (hunk 6 above)
Comment (petersutter) [reply frontend/src/components/ShootDetails/ShootCustomFieldsCard.vue]: > I could be better to pass the custom fields as property.

✅ also renamed the component to `CustomFieldsCard`

File: frontend/src/router/guards.js (hunk 8 above)
Comment (petersutter) [reply frontend/src/router/guards.js:138]: project/core/shoot/cli-gcp/terminal
projects/shoot-editor/warning

project/core/shoot-list/options
projects/shoot-list/options

project/core/shoot-list/selected-columns
projects/shoot-list/selected-columns

project/_all/shoot-list/filter

File: frontend/src/views/Administration.vue (hunk 15 above)
Comment (petersutter) [reply frontend/src/views/Administration.vue:152]: will rename to `shootCustomFields` and `shootCustomFieldList` respectively

File: frontend/src/components/CustomField.vue (hunk 1 above)
Comment (petersutter) [reply frontend/src/components/CustomField.vue:124]: ✅

File: frontend/src/components/CustomField.vue (hunk 3 above)
Comment (petersutter) [reply frontend/src/components/CustomField.vue:128]: added the required props (name, path). Don't think any default values need to be set, or do you have something in mind?

File: frontend/src/components/ShootListRow.vue (hunk 7 above)
Comment (petersutter) [reply frontend/src/components/ShootListRow.vue:11]: we agreed to discuss this topic again together with @grolu

File: frontend/src/store/index.js (hunk 13 above)
Comment (petersutter) [reply frontend/src/store/index.js:562]: ✅

File: frontend/src/views/Administration.vue (hunk 15 above)
Comment (petersutter) [reply frontend/src/views/Administration.vue:152]: ✅

File: frontend/src/router/guards.js (hunk 8 above)
Comment (petersutter) [reply frontend/src/router/guards.js:138]: I have adapted all keys, except for the terminals. Let's talk about that

File: frontend/src/store/index.js [hunk 26]
```diff
@@ -532,48 +532,50 @@ const getters = {
...
         key
       }
     })
   },
-  customFieldsShoot (state, getters) {
-    let customFieldsShoot = get(getters.projectFromProjectList, 'metadata.annotations["dashboard.gardener.cloud/customFieldsShoot"]')
-    if (!customFieldsShoot) {
-      return undefined
+  shootCustomFields (state, getters) {
+    function someObjectValuesAreObject () {
```
Comment (holgerkoser) [inline frontend/src/store/index.js:544]: Why is this a function factory? It has no parameters?

File: frontend/src/store/index.js [hunk 27]
```diff
@@ -532,48 +532,50 @@ const getters = {
...
+      columnSelectedByDefault: true,
+      showDetails: true,
+      sortable: true,
+      searchable: true
+    }
+    shootCustomFields = mapKeys(shootCustomFields, (customFields, key) => `Z_${key}`)
+    shootCustomFields = mapValues(shootCustomFields, customFields => {
+      return {
+        ...defaultProperties,
+        ...customFields
```
Comment (holgerkoser) [inline frontend/src/store/index.js:575]:
```suggestion
    shootCustomFields = mapKeys(shootCustomFields, (customField, key) => `Z_${key}`)
    shootCustomFields = mapValues(shootCustomFields, customField => {
      return {
        ...defaultProperties,
        ...customField
```

File: frontend/src/store/index.js [hunk 28]
```diff
@@ -532,48 +532,50 @@ const getters = {
...
+      return
     }
 
-    customFieldsShoot = pickBy(customFieldsShoot, customFields => {
-      return !isEmpty(get(customFields, 'path')) &&
-        !isEmpty(get(customFields, 'name'))
-    })
+    shootCustomFields = omitBy(shootCustomFields, isEmpty) // omit null values
+    shootCustomFields = omitBy(shootCustomFields, someObjectValuesAreObject())
+    shootCustomFields = pickBy(shootCustomFields, ({ path, name }) => name && path)
```
Comment (holgerkoser) [inline frontend/src/store/index.js:562]:
```suggestion
    shootCustomFields = filter(shootCustomFields, customField => {
      if (isEmpty(customField) {
        return false // omit null values
      }
      if (some(customField, isObject)) {
        return false // omit custom fields with object values
      }
      return customField.name && customField.path
    }) 
```

Comment (petersutter) [reply frontend/src/store/index.js:562]: filter returns an array, but the object should be preserved

File: frontend/src/store/index.js (hunk 27 above)
Comment (petersutter) [reply frontend/src/store/index.js:575]: ✅

File: frontend/src/store/index.js (hunk 26 above)
Comment (petersutter) [reply frontend/src/store/index.js:544]: ✅

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Review by grolu — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
👏 nice!

**Comment by HerrmannHinz** [comment]
nice one indeed! exactly what i was looking for! 🥳
